libuavcan/include/libuavcan/time.hpp

```diff
@@ -14,153 +15,252 @@
+        return fromMicrosecond(std::numeric_limits<USecT>::max());
     }
 
-    static DurationType fromMicrosecond(std::int64_t us)
+    static Type fromMicrosecond(USecT us)
     {
-        DurationType d;
+        Type d;
         d.usec_ = us;
         return d;
     }
 
-    static DurationType fromMillisecond(std::int64_t ms)
+    USecT toMicrosecond() const
     {
-        return fromMicrosecond(ms * 1000);
+        return usec_;
     }
 
-    std::int64_t toMicrosecond() const
+    USecT toMillisecond() const
     {
-        return usec_;
+        return usec_ / static_cast<USecT>(1000);
+    }
+
+    Type getAbs() const
+    {
+        return Type::fromMicrosecond(std::abs(usec_));
+    }
+
+    bool operator==(const Type& r) const
+    {
+        return usec_ == r.usec_;
+    }
+
+    bool operator!=(const Type& r) const
+    {
+        return !operator==(r);
     }
 
-    std::int64_t toMillisecond() const
+    bool operator<(const Type& r) const
     {
-        return usec_ / 1000;
+        return usec_ < r.usec_;
     }
 
-    DurationType getAbs() const
+    bool operator>(const Type& r) const
     {
-        return DurationType::fromMicrosecond((usec_ < 0) ? (-usec_) : usec_);
+        return usec_ > r.usec_;
     }
 
-    bool isPositive() const
+    bool operator<=(const Type& r) const
     {
-        return usec_ > 0;
+        return usec_ <= r.usec_;
     }
 
-    bool isNegative() const
+    bool operator>=(const Type& r) const
     {
-        return usec_ < 0;
+        return usec_ >= r.usec_;
     }
 
-    bool isZero() const
+    Type operator+(const Type& r) const
     {
-        return usec_ == 0;
+        return fromMicrosecond(util::saturating_add(usec_, r.usec_));
     }
 
-    bool operator==(const DurationType& r) const
+    Type operator-(const Type& r) const
+    {
+        return fromMicrosecond(util::saturating_sub(usec_, r.usec_));
+    }
+
+    Type operator-() const
+    {
+        return fromMicrosecond(-usec_);
+    }
+
+    Type& operator+=(const Type& r)
+    {
+        *this = *this + r;
+        return *static_cast<Type*>(this);
+    }
+
+    Type& operator-=(const Type& r)
+    {
+        *this = *this - r;
+        return *static_cast<Type*>(this);
+    }
+};
+
+class LIBUAVCAN_EXPORT Monotonic : public Base<Monotonic>
+{};
+
+}  // namespace duration
+
+namespace time
+{
+/**
+ * Protected base class for time values.
+ *
+ * @tparam Type          The type of the derived class. This must be an empty type.
+ *                       All storage will be provided by this base class.
+ * @tparam DType         The type of duration used for this time type. Time is concrete and duration
+ *                       is relative.
+ * @tparam USecT         The datatype returned when retrieving time from
+ *                       realizations of this base class. This type must be unsigned.
+ */
+template <typename Type, typename DType, typename USecT = uint64_t>
+class Base
+{
+    USecT usec_;
```

Comment:
```suggestion
    USecT usec_ = 0;
```

HIC++ 12.4.2, C++CG C.45

Reply:
See comment below: will define all defaults.